<a 
 href="https://colab.research.google.com/github/LearnPythonWithRune/FinancialDataAnalysisWithPython/blob/main/colab/starter/05 - Pandas Datareader.ipynb"
 target="_parent">
<img 
 src="https://colab.research.google.com/assets/colab-badge.svg"
alt="Open In Colab"/>
</a>

# 5. Pandas-datareader read data with web service API
- https://pandas-datareader.readthedocs.io/en/latest/

In Financial analysis, we need a lot fo historical data. These data are generally provided by many online financial serivces like Nasdaq, Woeld bank and goveronment institutions and commercial serivces. 

These serivces are normally provide data with a service API. Pandas needs to read data form these APIs. Pandas provides a function called **pandas-datareader** it can read data dirctly from a serivce provider. The detailed document can be found from [pandas-datareader](https://pandas-datareader.readthedocs.io/en/latest/)

Functions from pandas_datareader.data and pandas_datareader.wb extract data from various Internet sources into a pandas DataFrame. Currently the following sources are supported:

- Tiingo - IEX - Alpha Vantage - Econdb - Enigma - Quandl - St.Louis FED (FRED) - Kenneth French’s data library

- World Bank - OECD - Eurostat - Thrift Savings Plan - Nasdaq Trader symbol definitions - Stooq

- MOEX - Naver Finance

- Yahoo Finance

In [ ]:
# Install using pip
#!pip install pandas-datareader
#!pip install git+https://github.com/pydata/pandas-datareader.git

In [28]:
import pandas_datareader as pdr

#This example reads 5-years of 10-year constant maturity yields on U.S. government bonds.
data = pdr.get_data_fred('GS10')

In [29]:
data.head()

GS10
DATE            
2018-03-01  2.84
2018-04-01  2.87
2018-05-01  2.98
2018-06-01  2.91
2018-07-01  2.89

In [30]:
data.tail()

GS10
DATE            
2022-09-01  3.52
2022-10-01  3.98
2022-11-01  3.89
2022-12-01  3.62
2023-01-01  3.53

In [ ]:
## Read data from Yahoo Finance 

In [31]:
import datetime as dt

In [51]:
# Ticker is the symbol of a company on the Stock market. AAPL is Apple
ticker = "AAPL"
start = dt.datetime(2019, 1, 1)
end = dt.datetime(2020, 12, 31)

data = pdr.get_data_yahoo(ticker, start, end)


TypeError: string indices must be integers

In [33]:
data.head()

AttributeError: 'YahooDailyReader' object has no attribute 'head'

In [38]:
df = pdr.DataReader('yahoo', start='2019-09-10', end='2019-10-09')

NotImplementedError: data_source=None is not implemented

## Alternative source

If yahoo finance does not work, we cna use alternative services. such as yfinance.

In [45]:
import yfinance as yf
ticker = "AAPL"
start_date = "2020-01-27"
end_date = "2021-01-27"
data = yf.download(ticker, start=start_date, end=end_date)

[*********************100%***********************]  1 of 1 completed


In [43]:
data.head()

Open       High        Low      Close  Adj Close     Volume
Date                                                                        
2020-01-27  77.514999  77.942497  76.220001  77.237503  75.552483  161940000
2020-01-28  78.150002  79.599998  78.047501  79.422501  77.689812  162234000
2020-01-29  81.112503  81.962502  80.345001  81.084999  79.316055  216229200
2020-01-30  80.135002  81.022499  79.687500  80.967499  79.201111  126743200
2020-01-31  80.232498  80.669998  77.072502  77.377502  75.689445  199588400

In [46]:
data.tail()

Open        High         Low       Close   Adj Close  \
Date                                                                     
2021-01-20  128.660004  132.490005  128.550003  132.029999  130.264984   
2021-01-21  133.800003  139.669998  133.589996  136.869995  135.040268   
2021-01-22  136.279999  139.850006  135.020004  139.070007  137.210861   
2021-01-25  143.070007  145.089996  136.539993  142.919998  141.009384   
2021-01-26  143.600006  144.300003  141.369995  143.160004  141.246170   

               Volume  
Date                   
2021-01-20  104319500  
2021-01-21  120150900  
2021-01-22  114459400  
2021-01-25  157611700  
2021-01-26   98390600

**The data is same with we download from the website and save into a CSV fiel and load the content of the CSV file in to DataFrame**

In [47]:
data.index

DatetimeIndex(['2020-01-27', '2020-01-28', '2020-01-29', '2020-01-30',
               '2020-01-31', '2020-02-03', '2020-02-04', '2020-02-05',
               '2020-02-06', '2020-02-07',
               ...
               '2021-01-12', '2021-01-13', '2021-01-14', '2021-01-15',
               '2021-01-19', '2021-01-20', '2021-01-21', '2021-01-22',
               '2021-01-25', '2021-01-26'],
              dtype='datetime64[ns]', name='Date', length=253, freq=None)

In [48]:
data.dtypes

Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume         int64
dtype: object

## Read historic stock prices data from Stooq

Another open source you can get historic stock prices is from Stooq. Stooq is a website that provides historical financial data including OHLCV (Open, High, Low, Close, Volume) data on over 21,000 global securities and ETFs, data on global indices, commodities, and bonds, 1980 currency pairs and 132 cryptocurrencies. the website is managed by a Polish firm, historical stock prices in US dollars can be downloaded with symbols familiar to US stock traders and analysts.

In [64]:
data2 = pdr.get_data_stooq(ticker, start, end)

In [68]:
# understand data we get. we can use .head(), .info(), .shape, .describe(), .columns, .index, .values
data2.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 505 entries, 2020-12-31 to 2019-01-02
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Open    505 non-null    float64
 1   High    505 non-null    float64
 2   Low     505 non-null    float64
 3   Close   505 non-null    float64
 4   Volume  505 non-null    float64
dtypes: float64(5)
memory usage: 23.7 KB


In [55]:
data2.head()

Open      High      Low   Close      Volume
Date                                                      
2023-02-13  150.952  154.2600  150.920  153.85  62199013.0
2023-02-10  149.460  151.3401  149.220  151.01  57450708.0
2023-02-09  153.775  154.3300  150.420  150.87  56007143.0
2023-02-08  153.880  154.5800  151.168  151.92  64120079.0
2023-02-07  150.640  155.2300  150.640  154.65  83322551.0

In [56]:
data2.tail()

Open     High      Low    Close        Volume
Date                                                        
2019-01-08  36.0899  36.6342  35.8367  36.3711  1.700222e+08
2019-01-07  35.8786  35.9104  35.2046  35.6911  2.270168e+08
2019-01-04  34.8746  35.8446  34.6991  35.7729  2.428866e+08
2019-01-03  34.7431  35.1588  34.2635  34.3094  3.784272e+08
2019-01-02  37.3770  38.3281  37.2125  38.1048  1.535046e+08

## Other useful functions from Pandas-datareader

There are many other functions provided by the Pandas-datareader. Most fo them needs a company ticker (symbol).
We can use get_nasdaq_symbols to find a comany's symbol.
- find Nasdaq symbols.

In [69]:
nasdaq_sym = pdr.get_nasdaq_symbols()

nasdaq_sym.head()

Nasdaq Traded                                      Security Name  \
Symbol                                                                     
A                True            Agilent Technologies, Inc. Common Stock   
AA               True                    Alcoa Corporation Common Stock    
AAA              True  Investment Managers Series Trust II AXS First ...   
AAAU             True             Goldman Sachs Physical Gold ETF Shares   
AAC              True  Ares Acquisition Corporation Class A Ordinary ...   

       Listing Exchange Market Category    ETF  Round Lot Size  Test Issue  \
Symbol                                                                       
A                     N                  False           100.0       False   
AA                    N                  False           100.0       False   
AAA                   P                   True           100.0       False   
AAAU                  Z                   True           100.0       False   
AAC                   N                  False           100.0       False   

       Financial Status CQS Symbol NASDAQ Symbol  NextShares  
Symbol                                                        
A                   NaN          A             A       False  
AA                  NaN         AA            AA       False  
AAA                 NaN        AAA           AAA       False  
AAAU                NaN       AAAU          AAAU       False  
AAC                 NaN        AAC           AAC       False

In [70]:
# shape returns the number of rows and columns of the DataFrame
nasdaq_sym.shape

(11717, 11)

In [71]:
# info() shows information on each of the columns
nasdaq_sym.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11717 entries, A to ZYXI
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   Nasdaq Traded     11717 non-null  bool    
 1   Security Name     11717 non-null  object  
 2   Listing Exchange  11717 non-null  category
 3   Market Category   11717 non-null  object  
 4   ETF               11717 non-null  bool    
 5   Round Lot Size    11717 non-null  float64 
 6   Test Issue        11717 non-null  bool    
 7   Financial Status  5426 non-null   category
 8   CQS Symbol        6291 non-null   object  
 9   NASDAQ Symbol     11716 non-null  object  
 10  NextShares        11717 non-null  bool    
dtypes: bool(4), category(2), float64(1), object(4)
memory usage: 876.3+ KB


In [78]:
# find company Apple's symbol using many str functions such as .str.contains("xxxx")

Name = nasdaq_sym["Security Name"].str.contains("Apple Inc") 
nasdaq_sym[Name]

Nasdaq Traded              Security Name Listing Exchange  \
Symbol                                                              
AAPL             True  Apple Inc. - Common Stock                Q   

       Market Category    ETF  Round Lot Size  Test Issue Financial Status  \
Symbol                                                                       
AAPL                 Q  False           100.0       False                N   

       CQS Symbol NASDAQ Symbol  NextShares  
Symbol                                       
AAPL          NaN          AAPL       False

In [79]:
# ticker of Apple
ticker = nasdaq_sym[Name]["NASDAQ Symbol"]
ticker

Symbol
AAPL    AAPL
Name: NASDAQ Symbol, dtype: object

In [73]:
nasdaq_sym.loc['AAPL']

Nasdaq Traded                            True
Security Name       Apple Inc. - Common Stock
Listing Exchange                            Q
Market Category                             Q
ETF                                     False
Round Lot Size                          100.0
Test Issue                              False
Financial Status                            N
CQS Symbol                                NaN
NASDAQ Symbol                            AAPL
NextShares                              False
Name: AAPL, dtype: object

# End